[5.6] Look up in the JSON fallback translation file

ChristopheB opened the pull request:
For a given translation key, no lookup is actually being made in the JSON file for the fallback language. This PR adds this behaviour.

Now the lookup order for a given key is:

1. in the JSON translation file
2. in the typical translation file
3. in the typical translation file for the fallback language
4. with this pull request in the **JSON translation file for the fallback language**

(I still have to commit some tests)

ChristopheB committed:
Refactor Illuminate\Translation\Translator

ChristopheB committed:
Look up in the JSON fallback translation file

themsaid commented:
This is intended, If no translation found it should return the exact string sent to `__()`

themsaid commented:
This PR is actually a breaking change, and Travis tests are failing.

ChristopheB commented:
Yes, sorry about that, give some time to fix this later today.

ChristopheB committed:
Fix tests for Translator@getFromJson

ChristopheB committed:
Add a test for Translator@getFromJson about fallback language

When the key cannot be found anywhere except in the JSON translation file of the fallback language.

ChristopheB commented:
Let me explain my use case. I only use JSON translation files on my projects, and I just expect the fallback language to work. Also there is no change of behaviour, if no translation is found it still returns the key. This PR just adds a final lookup step in the JSON translation file for the fallback language.

(Tests pass now, sorry for the mess)

ChristopheB commented:
I ended up using a custom helper, something like this:

```php
function _($key = null, $replace = [], $locale = null) {
    return $key === ($translation = trans()->getFromJson($key, $replace, $locale)) ?
        trans()->getFromJson($key, $replace, trans()->getFallback()) :
        $translation;
}
```

ahmad-saad commented:
Hallo @taylorotwell ,

I have the same issue like @ChristopheB ,

I use the json files and now the fallback lang not working.

can you please explain why Laravel not looking inside the json File of the fallback lang.

Best Regard's

36864 commented:
When using translation strings, the language you use to look up a match IS the fallback language. There is no need to look up in a fallback file because the fallback string is already there.

ahmad-saad commented:
Hallo @36864 ,

ja you have right, but in other case this is not flexible.

let us say that we have the Langs (en, de, it, .....) and I create the translation by the json file.

en => is the fallback lang

for example I use {{ __('click hier to go to login page') }} then I add the translate to all the other langs files.

after a while I want to change the englisch version let us say to {{ __('go to login page') }}

now I must change all the keys in all files from 'click hier to go to login page' to ''go to login page'

to get the correct translation from the other langs.

so I must change all the other json files even if there no change in the translation there.

what you think about that??

Regard's

ChristopheB commented:
I've been using the custom helper described above with success on a large project, you should give it a try.
